packages/appium/lib/appium.js

```diff
@@ -241,6 +252,330 @@ class AppiumDriver extends DriverCore {
+    }
+
+    // Let's figure out which driver is going to handle this socket connection. It's either going
+    // to be a driver matching a session id appended to the bidi base path, or this umbrella driver
+    // (if no session id is included in the bidi connection request)
+
+    /** @type {import('@appium/types').ExternalDriver | AppiumDriver} */
+    let bidiHandlerDriver;
+
+    /** @type {import('ws').WebSocket | null} */
+    let proxyClient = null;
+
+    if (sessionMatch) {
+      // If we found a session id, see if it matches an active session
+      const sessionId = sessionMatch[1];
+      bidiHandlerDriver = this.sessions[sessionId];
+      // @ts-ignore this is actually assigned in the guard block above but ts doesn't recognize it
+      if (!bidiHandlerDriver) {
+        // The session ID sent in doesn't match an active session; just ignore this socket
+        // connection in that case
+        throw new Error(
+          `Got bidi connection request for session with id ${sessionId} but this is not an active ` +
+            `session. Closing the socket connection.`,
+        );
+      }
+      outOfBandErrorPrefix = `[session ${sessionId}] `;
+      this.log.info(`Bidi websocket connection made for session ${sessionId}`);
+      // store this socket connection for later removal on session deletion. theoretically there
+      // can be multiple sockets per session
+      if (!this.bidiSockets[sessionId]) {
+        this.bidiSockets[sessionId] = [];
+      }
+      this.bidiSockets[sessionId].push(ws);
+
+      const bidiProxyUrl = bidiHandlerDriver.bidiProxyUrl;
+      if (bidiProxyUrl) {
+        if (!_.isString(bidiProxyUrl)) {
+          throw new Error(
+            `Got request to proxy bidi connections to upstream socket with url ` +
+              `${bidiProxyUrl}, but this was not a valid url`,
+          );
+        }
+        this.log.info(`Bidi connection will be proxied to ${bidiProxyUrl}`);
+        proxyClient = new WebSocket(bidiProxyUrl);
+        this.bidiProxyClients[sessionId] = proxyClient;
+      }
+    } else {
+      this.log.info('Bidi websocket connection made to main server');
+      // no need to store the socket connection if it's to the main server since it will just
+      // stay open as long as the server itself is and will close when the server closes.
+      bidiHandlerDriver = this; // eslint-disable-line @typescript-eslint/no-this-alias
+    }
+
+    const logSocketErr = (/** @type {Error} */ err) =>
+      this.log.error(`${outOfBandErrorPrefix}${err}`);
+
+    // This is a function which wraps the 'send' method on a web socket for two reasons:
+    // 1. Make it async-await friendly
+    // 2. Do some logging if there's a send error
+    const sendFactory = (/** @type {import('ws').WebSocket} */ socket) => {
+      const socketSend = B.promisify(socket.send, {context: socket});
+      return async (/** @type {string|Buffer} */ data) => {
+        const err = await socketSend(data);
```

Comment:
I am confused. should it be try/catch instead?

Reply:
oops yeah refactor bug. good catch.